ranger.py

```diff
@@ -149,3 +174,107 @@ def __init__(self, level):
+        self.main_hand = main_hand
+        self.name = "BeastMasterAction"
+
+    def action(self, target: Target):
+        has_action = cast_spell(self.character, None)
+        if has_action:
+            if self.character.level >= 3:
+                if self.character.use_bonus("beast"):
+                    log.output(lambda: "Use bonus action for beast attack")
+                    self.beast.attack(target, self.maul)
+                    if self.character.level >= 11:
+                        self.beast.attack(target, self.maul)
+                    self.character.attack(target, self.main_hand)
+                else:
+                    if self.character.level >= 5:
+                        self.beast.attack(target, self.maul)
+                        if self.character.level >= 11:
+                            self.beast.attack(target, self.maul)
+                self.character.attack(target, self.main_hand)
+                self.character.attack(target, self.off_hand_nick)
+            else:
+                self.character.attack(target, self.main_hand)
+                if self.character.use_bonus("light weapon"):
+                    self.character.attack(target, self.off_hand_other)
+                else:
+                    self.character.attack(target, self.off_hand_nick)
+
+
+class BeastMasterRanger(Character):
+    def __init__(self, level):
+        self.magic_weapon = get_magic_weapon(level)
+        base_feats = []
+        maul = BeastMaul(base=2 + prof_bonus(level))
+        beast = PrimalCompanion(level, self, attack=maul)
+        shortsword = Shortsword(bonus=self.magic_weapon)
+        rapier = Rapier(bonus = self.magic_weapon)
+        other_shortsword = Shortsword(
+            bonus=self.magic_weapon,
+            name="Offhand Shortsword",
+            base="dex" if level >= 2 else 0,
+        )
+        scimitar = Scimitar(bonus=self.magic_weapon, base="dex" if level >= 2 else 0)
+        base_feats.append(EquipWeapon(maul))
+        base_feats.append(EquipWeapon(shortsword))
+        base_feats.append(EquipWeapon(rapier))
+        base_feats.append(EquipWeapon(scimitar))
+        base_feats.append(EquipWeapon(other_shortsword))
+        base_feats.append(BeastMasterAction(
+            beast=beast,
+            maul=maul,
+            main_hand=shortsword if level < 4 else rapier,
+            off_hand_nick=scimitar,
+            off_hand_other=other_shortsword,
+        ))
+        base_feats.append(Spellcasting(level, half=True))
+        base_feats.append(BeastChargeFeat(character=self))
+
+        def beast_only_level_11(args: HitArgs) -> bool:
+            return (not isinstance(args.attack.weapon, BeastMaul)) or (level >= 11)
+
+        if level >= 20:
+            base_feats.append(HuntersMarkFeat(10, True, ))
+        elif level >= 17:
+            base_feats.append(HuntersMarkFeat(6, True, ))
+        else:
+            base_feats.append(HuntersMarkFeat(6, False, filter=beast_only_level_11))
```

Comment:
Do you need this filter if the beast is a separate character? That character does not have the hunters mark feat

Reply:
Oh wait, beast master benefits from hunters mark when the beast hits the target. If that is the case, then you may be able to create a separate hunters mark feat and only add to the beast at level 11